libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/ptx/parallel_synchronization_and_communication_instructions_mbarrier.h

```diff
@@ -0,0 +1,392 @@
+
+// 9.7.12.15.12. Parallel Synchronization and Communication Instructions: mbarrier.complete_tx
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-complete-tx
+
+// 9.7.12.15.13. Parallel Synchronization and Communication Instructions: mbarrier.arrive
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-arrive
+
+// mbarrier.arrive{.sem}{.scope}{.shared{::cta}}.b64           state, [addr]{, count};
+// mbarrier.arrive{.sem}{.scope}{.shared::cluster}.b64         _, [addr] {,count}
+// mbarrier.arrive.expect_tx{.sem}{.scope}{.shared{::cta}}.b64 state, [addr], txCount;
+// mbarrier.arrive.expect_tx{.sem}{.scope}{.shared::cluster}.b64   _, [addr], txCount;
+// mbarrier.arrive.noComplete{.sem}{.cta}{.shared{::cta}}.b64  state, [addr], count;
+//
+// .sem   = { .release }
+// .scope = { .cta, .cluster }
+
+/*
+// mbarrier.arrive.shared.b64 state, [addr]; // 1.  PTX ISA 70, SM_80
+template <typename=void>
+__device__ static inline uint64_t mbarrier_arrive(
+  cuda::ptx::sem_release_t,
+  cuda::ptx::scope_cta_t,
+  cuda::ptx::space_shared_t,
+  uint64_t* addr);
+*/
+#if __cccl_ptx_isa >= 700
+extern "C" __device__ _CUDA_VSTD::uint64_t ___cuda_vstd_uint64_t__cuda_ptx_mbarrier_arrive_is_not_supported_before_SM_80__();
+template <typename=void>
+_LIBCUDACXX_DEVICE static inline _CUDA_VSTD::uint64_t mbarrier_arrive(
+  sem_release_t,
+  scope_cta_t,
+  space_shared_t,
+  _CUDA_VSTD::uint64_t* __addr)
+{
+  // __sem == sem_release (due to parameter type constraint)
+  // __scope == scope_cta (due to parameter type constraint)
+  // __space == space_shared (due to parameter type constraint)
+
+  _CUDA_VSTD::uint64_t __state;
+
+  NV_IF_ELSE_TARGET(NV_PROVIDES_SM_80,(
+    asm (
+      "mbarrier.arrive.shared.b64 %0, [%1]; // 1. "
+      : "=l"(__state)
+      : "r"(__as_ptr_smem(__addr))
+      : "memory"
+    );
+    return __state;
+  ),(
+    // Unsupported architectures will have a linker error with a semi-decent error message
+    return ___cuda_vstd_uint64_t__cuda_ptx_mbarrier_arrive_is_not_supported_before_SM_80__();
+  ));
+}
+#endif // __cccl_ptx_isa >= 700
+
+/*
+// mbarrier.arrive.noComplete.shared.b64 state, [addr], count; // 2.  PTX ISA 70, SM_80
+template <typename=void>
+__device__ static inline uint64_t mbarrier_arrive_no_complete(
+  cuda::ptx::sem_release_t,
+  cuda::ptx::scope_cta_t,
+  cuda::ptx::space_shared_t,
+  uint64_t* addr,
+  const uint32_t& count);
+*/
+#if __cccl_ptx_isa >= 700
+extern "C" __device__ _CUDA_VSTD::uint64_t ___cuda_vstd_uint64_t__cuda_ptx_mbarrier_arrive_no_complete_is_not_supported_before_SM_80__();
+template <typename=void>
+_LIBCUDACXX_DEVICE static inline _CUDA_VSTD::uint64_t mbarrier_arrive_no_complete(
+  sem_release_t,
+  scope_cta_t,
+  space_shared_t,
+  _CUDA_VSTD::uint64_t* __addr,
+  const _CUDA_VSTD::uint32_t& __count)
```

Comment:
question (non-blocking): Maybe this was discussed elsewhere, but what's the story on taking this parameter by `const&` instead of by value? I always get nervous when I see things like that because I've seen cases where taking parameters by ref in device functions can lead to local memory usage because nvcc has to reify the object in addressable memory (instead of a register.

Reply:
It was suggested by Mark Hoemmen. Apparently this has been observed to lead to better codegen in CUTLASS. My own experience is that for variables that are already in register, I have not observed any local memory usage due to using const references (also with fixed length arrays).